wlauto/utils/statedetect.py

```diff
@@ -0,0 +1,147 @@
+    lower = int(max(0, (1.0 - sigma) * v))
+    upper = int(min(255, (1.0 + sigma) * v))
+    edged = cv2.Canny(image, lower, upper)
+
+    # return the edged image
+    return edged
+
+
+def match_state(screenshot_file, defpath):
+    # check dependencies
+    if np == None or cv2 == None or imutils == None:
+        raise RuntimeError("State detection requires numpy, opencv (cv2) and imutils.")
+
+    # load and parse state definition file
+    statedefs_file = os.path.join(defpath, 'definition.yaml')
+    if not os.path.isfile(statedefs_file):
+        raise StateDefinitionError("Missing state definitions yaml file: "+statedefs_file)
+    with open(statedefs_file) as fh:
+        state_definitions = yaml.load(fh)
+
+    # check if file exists, then load screenshot into opencv and create edge map
+    if not os.path.isfile(screenshot_file):
+        raise StateDefinitionError("Screenshot file not found")
+    img_rgb = cv2.imread(screenshot_file)
+    img_gray = cv2.cvtColor(img_rgb, cv2.COLOR_BGR2GRAY)
+    img_edge = auto_canny(img_gray)
+
+    # make a list of all templates defined in the state definitions
+    template_list = []
+    for state in state_definitions["workload_states"]:
+        template_list.extend(state["templates"])
+
+    # check all template PNGs exist
+    missing_files = 0
+    for template_png in template_list:
+        if not os.path.isfile(os.path.join(defpath, 'templates', template_png+'.png')):
+            missing_files += 1
+
+    if missing_files:
+        raise StateDefinitionError("Missing template PNG files")
+
+    # try to match each PNG
+    matched_templates = []
+    for template_png in template_list:
+        template = cv2.imread(os.path.join(defpath, 'templates', template_png+'.png'), 0)
+        template_edge = auto_canny(template)
+        w, h = template.shape[::-1]
+
+        res = cv2.matchTemplate(img_edge, template_edge, cv2.TM_CCOEFF_NORMED)
+        threshold = 0.5
+        loc = np.where(res >= threshold)
+        zipped = zip(*loc[::-1])
+
+        if len(zipped) > 0:
+            matched_templates.append(template_png)
+
+    # determine the state according to the matched templates
+    matched_state = "none"
+    for state in state_definitions["workload_states"]:
+        # look in the matched templates list for each template of this state
+        matchCount = 0
+        for template in state["templates"]:
+            if template in matched_templates:
+                matchCount += 1
+
+        if matchCount >= state["matches"]:
+            # we have a match
+            matched_state = state["state_name"]
+            break
+
+    return matched_state
+
+
+def verify_state(screenshot_file, stateDefsPath, workloadPhase):
+    # run a match on the screenshot
+    matched_state = match_state(screenshot_file, stateDefsPath)
+
+    # load and parse state definition file
+    statedefs_file = os.path.join(stateDefsPath, 'definition.yaml')
```

Comment:
Rather than parsing the yaml file both in `match_state` and here, just parse it here and pass the parsed yaml to `match_state`